android/src/main/java/com/twiliorn/FrameCapturerVideoSink.java

```diff
@@ -0,0 +1,48 @@
+package com.twiliorn.library;
+
+import android.util.Log;
+
+import com.facebook.react.bridge.ReactContext;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import tvi.webrtc.VideoFrame;
+import tvi.webrtc.VideoSink;
+
+public class FrameCapturerVideoSink implements VideoSink {
+    private final AtomicBoolean captureThisFrame = new AtomicBoolean(false);
+    // TODO: FIX THIS. ITS A POTENTIAL MEMORY LEAK.
+    private ReactContext context;
+    // choosing to use single thread since limited i/o resources may be a bottleneck anyways
+    private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+    private String filename = "";
+
+    @Override
+    public void onFrame(VideoFrame frame) {
+        if (frame == null) {
+            Log.w(TwilioPackage.TAG, "Cannot capture null frame.");
+            return;
+        }
+
+        if (captureThisFrame.compareAndSet(true, false)) {
+            frame.retain(); // retain frame so we can save it on background thread, bg thread will handle releasing
+            Log.d(TwilioPackage.TAG, "Capturing frame on background thread.");
+            // save frame on background thread
+            executorService.execute(() -> {
+                Utils.saveVideoFrame(frame, context, this.filename);
```

Comment:
I didn't test this yet, but it looks like maybe a subtle bug lurking here.  What would happen if they clicked on the Streemshot button a few times quickly?  It seems like in captureFrame() it will overwrite this.filename with the new filename each time, and then when each of the queued threads starts up to execute Utils.saveVideoFrame(), they would all see the latest version of this.filename.  So for example if they clicked the Streemshot button 3 times, they might only end up with one Streemshot file (because the 2nd two requests just overwrite the same filename).  That might then also cause errors in the upload manger with those files not found.  

Possible fix would be to capture the current value of this.filename in a temp local variable that gets passed into this lambda?

Reply:
Great catch! I think you are right. I'll implement your recommended fix and do some testing around this edge case

Reply:
Ok, finally got some time to fix and test this. Took some streemshots as fast as web would allow and they were all able to saved without any overriding.